uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php

```diff
@@ -861,32 +859,73 @@ public static function findPublicHolidayLeaveRequestEvenIfDeleted($contactID, Pu
-   * Returns the leave request days already existing for the given contact
-   * within the $fromDate to $todDate period and having the statuses supplied.
+   * Returns existing leave requests considered "overlapping"
+   * for the given leave request parameters and statuses supplied.
    *
-   * @param int $contactID
-   * @param string $fromDate
-   * @param string $toDate
+   * @param array $leaveRequestParams
+   *   List of parameters to use to create the Case.
+   *   [
+   *     'contact_id' => int,
+   *     'type_id' => int,
+   *     'request_type' => int,
+   *     'from_date' => string,
+   *     'from_date_type' => string|int optional,
+   *     'to_date' => string,
+   *     'to_date_type' => string|int optional
+   *   ]
    * @param array $leaveRequestStatus
    * @param boolean $excludePublicHolidayLeaveRequests
    *   Whether to exclude public holiday leave requests from overlapping leave requests or not
    *
    * @return \CRM_HRLeaveAndAbsences_BAO_LeaveRequest[]|null
    */
-  public static function findOverlappingLeaveRequests($contactID, $fromDate, $toDate, $leaveRequestStatus = [], $excludePublicHolidayLeaveRequests = true) {
+  public static function findOverlappingLeaveRequests($leaveRequestParams, $leaveRequestStatus = [], $excludePublicHolidayLeaveRequests = TRUE) {
+    $fromDate = $leaveRequestParams['from_date'];
+    $fromDateType = isset($leaveRequestParams['from_date_type']) ? (int)$leaveRequestParams['from_date_type'] : NULL;
+    $toDate = $leaveRequestParams['to_date'];
+    $toDateType = isset($leaveRequestParams['to_date_type']) ? (int)$leaveRequestParams['to_date_type'] : NULL;
+
     $leaveRequestTable = self::getTableName();
     $leaveRequestDateTable = LeaveRequestDate::getTableName();
     $leaveBalanceChangeTable = LeaveBalanceChange::getTableName();
+    $leaveRequestAbsenceType = AbsenceType::findById($leaveRequestParams['type_id']);
+    $isCalculationUnitInHours = $leaveRequestAbsenceType->isCalculationUnitInHours();
+    $absenceTypes = AbsenceType::getEnabledAbsenceTypes();
+    $absenceTypesIDsWithSameUnit = [];
+    $absenceTypesIDsWithDifferentUnit = [];
+    $leaveRequestDayTypes = array_flip(self::buildOptions('from_date_type', 'validate'));
+    $halfDayAMID = $leaveRequestDayTypes['half_day_am'];
+    $halfDayPMID = $leaveRequestDayTypes['half_day_pm'];
+    $isTOIL = $leaveRequestParams['request_type'] === LeaveRequest::REQUEST_TYPE_TOIL;
+
+    foreach ($absenceTypes as $absenceType) {
+      if ($leaveRequestAbsenceType->calculation_unit === $absenceType->calculation_unit) {
+        $absenceTypesIDsWithSameUnit[] = $absenceType->id;
+      } else {
+        $absenceTypesIDsWithDifferentUnit[] = $absenceType->id;
+      }
+    }
+
+    $noOverlappingCondition =
+      $isCalculationUnitInHours || $isTOIL
+      ?
+        "(lr.from_date < '{$toDate}' AND lr.to_date > '{$fromDate}'" . (!$isTOIL ? " AND lr.type_id IN (" . join(',', $absenceTypesIDsWithSameUnit) . ")" : "") . ")" .
```

Comment:
These whole logic/query is very unreadable, It would be a good idea to separate them into smaller chunks and assign a readable variable name to each of them.

Reply:
I think done.